### PR TITLE
fix: Improve payload typings for Tasker handlers

### LINKED ISSUE #14979: [CAL-3683] Improve payload typings for Tasker handlers
Right now all payload just receive a string is used to stringify and parse JSON. We already use zod to parse on runtime. We should use z.infer to get that benefit on typeland.

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-3683](https://linear.app/calcom/issue/CAL-3683/improve-payload-typings-for-tasker-handlers)</sub>

### PR DESCRIPTION
**Pull request by anikdhabal**
## What does this PR do?

- Fixes #14979 

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [ ] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: packages/features/tasker/tasks/sendEmail.ts [hunk 1]
```diff
@@ -8,9 +8,12 @@ const sendEmailPayloadSchema = z.object({
   payload: z.string(),
 });
 
+type SendEmailPayload = z.infer<typeof sendEmailPayloadSchema>;
+
 export async function sendEmail(payload: string): Promise<void> {
   try {
-    const parsedPayload = sendEmailPayloadSchema.parse(JSON.parse(payload));
+    const parsedPayloadObj: SendEmailPayload = JSON.parse(payload);
```
Comment (utkershrajvenshi) [inline packages/features/tasker/tasks/sendEmail.ts:15]: So now if the string payload we pass to this function doesn't match the `SendEmailPayload` type that we have inferred from the zod schema defined earlier, we would get a `TypeError` exception, correct?

@anikdhabal Please help me out in understanding this aspect of the changes

File: packages/features/tasker/tasks/sendEmail.ts [hunk 2]
```diff
@@ -8,9 +8,12 @@ const sendEmailPayloadSchema = z.object({
...
+    const parsedPayload = sendEmailPayloadSchema.parse(parsedPayloadObj);
     console.log(parsedPayload);
```
Comment (utkershrajvenshi) [inline packages/features/tasker/tasks/sendEmail.ts:17]: Also, is this console.log intentional or can be removed safely?

File: packages/features/tasker/tasks/sendEmail.ts (hunk 1 above)
Comment (zomars) [reply packages/features/tasker/tasks/sendEmail.ts:15]: Payload is saved in DB as string always. So no. This won't prevent getting an error on compile time. We also need to keep the runtime protection as well.

**Review by zomars — CHANGES_REQUESTED** [review]
Thank you for your contribution 🙏 Left some comments about the approach of this PR.

The idea overall is close to what we want but in the current state is not giving us the compile time protection while removing the runtime one. Which is not an acceptable trade off.

I would recommend coming up with an approach that checks both compile time and runtime.

**Comment by anikdhabal** [comment]
> Thank you for your contribution 🙏 Left some comments about the approach of this PR.
> 
> The idea overall is close to what we want but in the current state is not giving us the compile time protection while removing the runtime one. Which is not an acceptable trade off.
> 
> I would recommend coming up with an approach that checks both compile time and runtime.

Thanks @zomars for your review. I will fix that shortly.

File: packages/features/tasker/tasks/sendEmail.ts (hunk 1 above)
Comment (utkershrajvenshi) [reply packages/features/tasker/tasks/sendEmail.ts:15]: Got it @zomars. Will have a further look

File: packages/features/tasker/tasks/sendEmail.ts [hunk 3]
```diff
@@ -8,9 +8,11 @@ const sendEmailPayloadSchema = z.object({
   payload: z.string(),
 });
 
+type SendEmailPayload = z.infer<typeof sendEmailPayloadSchema>;
+
 export async function sendEmail(payload: string): Promise<void> {
   try {
-    const parsedPayload = sendEmailPayloadSchema.parse(JSON.parse(payload));
+    const parsedPayload: SendEmailPayload = sendEmailPayloadSchema.parse(JSON.parse(payload));
```
Comment (zomars) [inline packages/features/tasker/tasks/sendEmail.ts:15]: Hey @anikdhabal as far as I know, parsedPayload already has the correct type inference. These typing should be added to the packages/features/tasker/tasker.ts typings in order to keep the task creation type safe. Not in here. Will close the PR for now. 🙏

**Review by zomars — CHANGES_REQUESTED** [review]

**Pull request closed without merging**